#! /bin/sh -e
# 
#
# DP: Author: Kazuhiro Inaoka
# DP: Fixed PR target/26775 target/26776

if [ $# -eq 3 ] && [ "$2" = '-d' ]; then
    pdir="-d $3"
    dir="$3/"
elif [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
       -patch) patch $pdir -f --no-backup-if-mismatch -p1 < $0;;
       -unpatch) patch $pdir -f --no-backup-if-mismatch -R -p1 < $0;;
	*)
		echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
		exit 1;;
esac

exit 0

--- gcc/gcc/config/m32r/m32r.c.orig	2006-03-01 16:28:44.000000000 +0900
+++ gcc/gcc/config/m32r/m32r.c	2006-03-21 15:33:53.894564698 +0900
@@ -2087,11 +2087,19 @@ m32r_output_function_epilogue (FILE * fi
 	  else if (reg_offset < 32768)
 	    fprintf (file, "\tadd3 %s,%s,%s%d\n",
 		     sp_str, sp_str, IMMEDIATE_PREFIX, reg_offset);
-	  else
+	  else if (reg_offset <= 0xffffff)
 	    fprintf (file, "\tld24 %s,%s%d\n\tadd %s,%s\n",
 		     reg_names[PROLOGUE_TMP_REGNUM],
 		     IMMEDIATE_PREFIX, reg_offset,
 		     sp_str, reg_names[PROLOGUE_TMP_REGNUM]);
+	  else
+	    fprintf (file, "\tseth %s,%shigh(%d)\n\tor3 %s,%s,%slow(%d)\n\tadd %s,%s\n",
+		     reg_names[PROLOGUE_TMP_REGNUM],
+		     IMMEDIATE_PREFIX, reg_offset,
+		     reg_names[PROLOGUE_TMP_REGNUM],
+		     reg_names[PROLOGUE_TMP_REGNUM],
+		     IMMEDIATE_PREFIX, reg_offset,
+		     sp_str, reg_names[PROLOGUE_TMP_REGNUM]);
 	}
       else if (frame_pointer_needed)
 	{
@@ -2102,10 +2110,20 @@ m32r_output_function_epilogue (FILE * fi
 	  else if (reg_offset < 32768)
 	    fprintf (file, "\tadd3 %s,%s,%s%d\n",
 		     sp_str, fp_str, IMMEDIATE_PREFIX, reg_offset);
+	  else if (reg_offset <= 0xffffff)
+	    fprintf (file, "\tld24 %s,%s%d\n\tadd %s,%s\n\tmv %s,%s\n",
+		     reg_names[PROLOGUE_TMP_REGNUM],
+		     IMMEDIATE_PREFIX, reg_offset,
+		     reg_names[PROLOGUE_TMP_REGNUM], fp_str,
+		     sp_str, reg_names[PROLOGUE_TMP_REGNUM]);
 	  else
-	    fprintf (file, "\tld24 %s,%s%d\n\tadd %s,%s\n",
+	    fprintf (file, "\tseth %s,%shigh(%d)\nor3 %s,%s,%slow(%d)\n\tadd %s,%s\n\tmv %s,%s\n",
+		     reg_names[PROLOGUE_TMP_REGNUM],
+		     IMMEDIATE_PREFIX, reg_offset,
+		     reg_names[PROLOGUE_TMP_REGNUM],
 		     reg_names[PROLOGUE_TMP_REGNUM],
 		     IMMEDIATE_PREFIX, reg_offset,
+		     reg_names[PROLOGUE_TMP_REGNUM], fp_str,
 		     sp_str, reg_names[PROLOGUE_TMP_REGNUM]);
 	}
       else
@@ -2127,8 +2145,26 @@ m32r_output_function_epilogue (FILE * fi
 
       /* Remove varargs area if present.  */
       if (current_frame_info.pretend_size != 0)
-	fprintf (file, "\taddi %s,%s%d\n",
-		 sp_str, IMMEDIATE_PREFIX, current_frame_info.pretend_size);
+        if (current_frame_info.pretend_size < 128)
+	  fprintf (file, "\taddi %s,%s%d\n",
+		     sp_str, IMMEDIATE_PREFIX, current_frame_info.pretend_size);
+        else if (current_frame_info.pretend_size < 32768)
+	  fprintf (file, "\tadd3 %s,%s,%s%d\n",
+		   sp_str, sp_str, IMMEDIATE_PREFIX,
+		   current_frame_info.pretend_size);
+        else if (current_frame_info.pretend_size <= 0xffffff)
+	  fprintf (file, "\tld24 %s,%s%d\n\tadd %s,%s\n",
+		   reg_names[PROLOGUE_TMP_REGNUM],
+		   IMMEDIATE_PREFIX, current_frame_info.pretend_size,
+		   sp_str, reg_names[PROLOGUE_TMP_REGNUM]);
+        else
+	  fprintf (file, "\tseth %s,%shigh(%d)\n\tor3 %s,%s,%slow(%d)\n\tadd %s,%s\n",
+		   reg_names[PROLOGUE_TMP_REGNUM],
+		   IMMEDIATE_PREFIX, current_frame_info.pretend_size,
+		   reg_names[PROLOGUE_TMP_REGNUM],
+		   reg_names[PROLOGUE_TMP_REGNUM],
+		   IMMEDIATE_PREFIX, current_frame_info.pretend_size,
+		   sp_str, reg_names[PROLOGUE_TMP_REGNUM]);
 	
       /* Emit the return instruction.  */
       if (M32R_INTERRUPT_P (fn_type))
